File: swift/Sources/WalletAPI/TransactionLog.swift
import Foundation

/// Audit log of wallet transactions — presentations and issuances (ARF transaction logging / GDPR).
/// A cross-cutting port: injected by the host. Production wallets MUST persist these; the default is no-op.
public protocol TransactionLog: Sendable {
    func record(_ entry: TransactionLogEntry) async throws
    func list() async throws -> [TransactionLogEntry]
}

public struct TransactionLogEntry: Sendable {
    public let id: String
    public let type: TransactionType
    public let timestamp: Date
    /// Relying party / verifier client_id (presentation) or issuer (issuance).
    public let relyingParty: String?
    public let credentialIds: [String]
    /// Disclosed claim paths, dot-joined (presentation only).
    public let claimsDisclosed: [String]
    public let status: TransactionStatus

    public init(id: String, type: TransactionType, timestamp: Date, relyingParty: String?,
                credentialIds: [String], claimsDisclosed: [String], status: TransactionStatus) {
        self.id = id
        self.type = type
        self.timestamp = timestamp
        self.relyingParty = relyingParty
        self.credentialIds = credentialIds
        self.claimsDisclosed = claimsDisclosed
        self.status = status
    }
}

public enum TransactionType: Sendable { case presentation, issuance }

public enum TransactionStatus: Sendable { case success, declined, failed }

/// Default no-op log. Replace with a persistent adapter for production audit.
public struct NoOpTransactionLog: TransactionLog {
    public init() {}
    public func record(_ entry: TransactionLogEntry) async throws {}
    public func list() async throws -> [TransactionLogEntry] { [] }
}

File: kotlin/wallet-api/src/main/kotlin/com/hopae/eudi/wallet/spi/TransactionLog.kt
package com.hopae.eudi.wallet.spi

import java.time.Instant

/**
 * Audit log of wallet transactions — presentations and issuances (ARF transaction logging / GDPR).
 * A cross-cutting port: injected by the host. Production wallets MUST persist these; the default is no-op.
 */
interface TransactionLog {
    suspend fun record(entry: TransactionLogEntry)
    suspend fun list(): List<TransactionLogEntry>
}

class TransactionLogEntry(
    val id: String,
    val type: TransactionType,
    val timestamp: Instant,
    /** Relying party / verifier client_id (presentation) or issuer (issuance). */
    val relyingParty: String?,
    val credentialIds: List<String>,
    /** Disclosed claim paths, dot-joined (presentation only). */
    val claimsDisclosed: List<String>,
    val status: TransactionStatus,
)

enum class TransactionType { Presentation, Issuance }

enum class TransactionStatus { Success, Declined, Failed }

/** Default no-op log. Replace with a persistent adapter for production audit. */
object NoOpTransactionLog : TransactionLog {
    override suspend fun record(entry: TransactionLogEntry) {}
    override suspend fun list(): List<TransactionLogEntry> = emptyList()
}
